src/Native/System.Security.Cryptography.Native/openssl.c

```diff
@@ -719,6 +719,224 @@ GetX509NameInfo(
+{
+    assert(candidate);
+    assert(hostname);
+
+    if (!candidate->data || !candidate->length)
+    {
+        return 0;
+    }
+
+    // If the candidate is *.example.org then the smallest we would match is a.example.org, which is the same
+    // length. So anything longer than what we're matching against isn't valid.
+
+    // This assumption might not hold under IDNA, it might need to be more sophisticated
+    // "UTF character string length"
+    if (candidate->length > cchHostname)
+    {
+        return 0;
+    }
+
+    if (sanRules)
+    {
+        // RFC2818 says to use RFC2595 matching rules, but then gives an example that f*.com would match foo.com
+        // RFC2595 says that '*' may be used as the left name component, in which case it is a wildcard that does
+        // not match a '.'.
+        //
+        // In the interest of time, and the idea that it's better to err on the side of more restrictive,
+        // this implementation does not support mid-string wildcards.
+        //
+        // TODO (3444): Determine if we're too restrictive here.
+
+        char* candidateStr;
+        int i;
+        int hostnameFirstDot = -1;
+
+        // A GEN_DNS name is supposed to be a V_ASN1_IA5STRING.  If it isn't, we don't know how to read it.
+        if (candidate->type != V_ASN1_IA5STRING)
+        {
+            return 0;
+        }
+
+        // Great, candidateStr is just candidate->data!
+        candidateStr = (char*)candidate->data;
+
+        // First, verify that the string is alphanumeric, plus hyphens or periods and maybe starting with an asterisk.
+        for (i = 0; i < candidate->length; ++i)
+        {
+            char c = candidateStr[i];
+
+            if ((c < 'a' || c > 'z') &&
+                (c < '0' || c > '9') &&
+                (c != '.') &&
+                (c != '-') &&
+                (c != '*' || i != 0))
+            {
+                return 0;
+            }
+        }
+
+        if (candidateStr[0] != '*')
+        {
+            if (candidate->length != cchHostname)
+            {
+                return 0;
+            }
+
+            return !memcmp(candidateStr, hostname, (size_t)cchHostname);
+        }
+
+        for (i = 0; i < cchHostname; ++i)
+        {
+            if (hostname[i] == '.')
+            {
+                hostnameFirstDot = i;
+                break;
+            }
+        }
+
+        if (hostnameFirstDot < 0)
+        {
+            // It's possible that this should be considered a match if the entire SAN entry is '*',
+            // aka candidate->length == 1; but nothing talks about this case.
+            return 0;
+        }
+
+        {
+            // If hostname is 10 bytes long, and the '.' was at index 3, then we eliminate the first 3
+            // characters (www) from the match constraint.  This forces the wildcard to be the last
+            // character before the . in its match group.
```

Comment:
That comment was mainly me explaining to myself that there wasn't an off-by-one error in the math.  If it's adding confusion I can delete it (or some rewording suggestions would be welcome).